File: scripts/ColorSubscriber.py
#!/usr/bin/python
import os
import rospy
from openhab_msgs.msg import ColorState

class ColorSubscriber(object):

    def __init__(self, item_name):
        self.item_name = item_name

        """Configure subscriber."""
        # Create a subscriber with appropriate topic, custom message and name of
        # callback function.
        self.sub = rospy.Subscriber("/openhab/items/" + self.item_name + "/state", ColorState, self.callback)

        # Initialize message variables.
        self.enable = False
        self.data = ""

        if self.enable:
            self.start()
        else:
            self.stop()

    def start(self):
        self.enable = True
        self.sub = rospy.Subscriber("/openhab/items/" + self.item_name + "/state", ColorState, self.callback)

    def callback(self, data):
        """Handle subscriber data."""
        # Simply print out values in our custom message.
        self.data = data
        msg = "Received %s with r %s, g %s and b %s" % (self.data.item, self.data.r, self.data.g, self.data.b)
        rospy.loginfo(rospy.get_caller_id() + msg)

    def stop(self):
        """Turn off subscriber."""
        self.enable = False
        self.sub.unregister()

if __name__ == '__main__':
    # Initialize the node and name it.
    node_name = "ColorSubscriberNode"
    rospy.init_node(node_name, anonymous=True)

    colorSubscriber = ColorSubscriber("testColor")

    # Go to the main loop
    try:
        colorSubscriber.start()
        # Wait for messages on topic, go to callback function when new messages arrive.
        # spin() simply keeps python from exiting until this node is stopped
        rospy.spin()
    # Stop with Ctrl + C
    except KeyboardInterrupt:
        colorSubscriber.stop()

        nodes = os.popen("rosnode list").readlines()
        for i in range(len(nodes)):
            nodes[i] = nodes[i].replace("\n","")

        for node in nodes:
            os.system("rosnode kill " + node_name)


        print("Node stopped")

File: src/ColorSubscriber.cpp
#include <ros/ros.h>
#include <openhab_msgs/ColorState.h>

class ColorSubscriber {
public:
    ColorSubscriber(std::string item_name) : item_name_(item_name) {
        // Create a subscriber with appropriate topic, custom message and name of
        // callback function.
        sub_ = nh_.subscribe("/openhab/items/" + item_name_ + "/state", 10, &ColorSubscriber::callback, this);

        // Initialize message variables.
        enable_ = false;
        data_ = NULL;

        if (enable_) {
            start();
        } else {
            stop();
        }
    }

    void start() {
        enable_ = true;
        sub_ = nh_.subscribe("/openhab/items/" + item_name_ + "/state", 10, &ColorSubscriber::callback, this);
    }

    void stop() {
        enable_ = false;
        sub_.shutdown();
    }

    void callback(const openhab_msgs::ColorState& data) {
        // Simply print out values in our custom message.
        data_ = data;
        if (!data_.isnull) {
            ROS_INFO("Received %s with hue %f, saturation %f and brightness %f", data_.item.c_str(), data_.hue, data_.saturation, data_.brightness);
        } else {
            ROS_INFO("Received %s with NULL", data_.item.c_str());
        }
    }

private:
    ros::NodeHandle nh_;
    ros::Subscriber sub_;
    std::string item_name_;
    bool enable_;
    openhab_msgs::ColorState data_;
};

int main(int argc, char** argv) {
    // Initialize the node and name it.
    std::string node_name = "ColorSubscriberNode";
    ros::init(argc, argv, node_name);

    ColorSubscriber colorSubscriber("testColor");

    // Wait for messages on topic, go to callback function when new messages arrive.
    // spin() simply keeps C++ from exiting until this node is stopped
    ros::spin();

    return 0;
}
